### PR TITLE
EventDispatcher Correctness and Documentation Improvements

### PR DESCRIPTION
**Pull request by simpsont-oci**
Problems:
 - Current EventDispatcher implementation (ServiceEventDispatcher / DispatchService) has a few issues regarding edge-case behavior: prohibiting null/empty events, exception handling around calling user event methods, event cancellation lock ordering.

Solution:
 - Improve edge-case checks, improve test coverage, improve documentation

### CODE REVIEW

File: dds/DCPS/DispatchService.h [hunk 1]
```diff
@@ -61,7 +69,19 @@ class OpenDDS_Dcps_Export DispatchService : public RcObject {
 
   /**
    * Request shutdown of this DispatchService, which prevents sucessful future
```
Comment (sonndinh) [inline dds/DCPS/DispatchService.h:71]:
```suggestion
   * Request shutdown of this DispatchService, which prevents successful future
```

File: dds/DCPS/EventDispatcher.h [hunk 2]
```diff
@@ -111,21 +120,27 @@ class OpenDDS_Dcps_Export EventDispatcher : public RcObject {
 
   /**
    * Request shutdown of this EventDispatcher, which prevents sucessful future
```
Comment (sonndinh) [inline dds/DCPS/EventDispatcher.h:122]:
```suggestion
   * Request shutdown of this EventDispatcher, which prevents successful future
```

File: dds/DCPS/ServiceEventDispatcher.cpp [hunk 3]
```diff
@@ -7,15 +7,37 @@
...
+{
+  if (!ptr) {
+    return;
+  }
+
+  try {
+    ptr->handle_cancel();
+  } catch (...) {
+    if (log_level >= LogLevel::Warning) {
+      ACE_ERROR((LM_WARNING, "(%P|%t) WARNING: ServiceEventDispatcher: handle_cancel threw an exception\n"));
```
Comment (sonndinh) [inline dds/DCPS/ServiceEventDispatcher.cpp:30]: Include the function name in the log message?

**Review by sonndinh — APPROVED** [review]
